File: common/src/main/java/io/wdsj/asw/common/type/ModuleType.java
package io.wdsj.asw.common.type;

import java.util.Arrays;
import java.util.List;

/**
 * Types for different detection modules.
 */
public enum ModuleType {
    CHAT(true),
    AI(true),
    SIGN(true),
    ANVIL(true),
    BOOK(true),
    NAME(false),
    ITEM(true),
    BROADCAST(false);

    private final boolean violationTracked;

    ModuleType(boolean violationTracked) {
        this.violationTracked = violationTracked;
    }

    public boolean isViolationTracked() {
        return violationTracked;
    }

    public static List<ModuleType> violationModules() {
        return Arrays.stream(values())
                .filter(ModuleType::isViolationTracked)
                .toList();
    }

    public static ModuleType parseViolationModule(String value) {
        if (value == null) {
            return null;
        }
        return Arrays.stream(values())
                .filter(ModuleType::isViolationTracked)
                .filter(moduleType -> moduleType.name().equalsIgnoreCase(value))
                .findFirst()
                .orElse(null);
    }
}

File: paper/src/main/kotlin/io/wdsj/asw/bukkit/type/ModuleType.kt
package io.wdsj.asw.bukkit.type

/**
 * Types for different detection modules.
 */
enum class ModuleType(val isViolationTracked: Boolean) {
    CHAT(true),
    AI(true),
    SIGN(true),
    ANVIL(true),
    BOOK(true),
    NAME(false),
    ITEM(true);

    companion object {
        @JvmStatic
        fun violationModules(): List<ModuleType> = entries.filter(ModuleType::isViolationTracked)

        @JvmStatic
        fun parseViolationModule(value: String): ModuleType? =
            entries.firstOrNull {
                it.isViolationTracked && it.name.equals(value, ignoreCase = true)
            }
    }
}
